### DISCUSSION TITLE
a DistributedExecutor equivalent to CeleryExecutor without Celery

### DISCUSSION THREAD

**Discussion by raphaelauv** [apache/airflow#26689] [Ideas]
Since the scheduler is capable of doing HA thanks to `SKIP LOCKED`

https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html#running-more-than-one-scheduler

a distributed airflow executor without celery ( and it's backend redis/rabbitmq ) would give the possibility to have a simpler design for a airflow distributed mode with workers ( the same way the triggerer(s) do not need celery )

wdyt ?

ps : there is an open discussion on celery about using postgres as a broker backend -> https://github.com/celery/celery/issues/5149

**Comment by potiuk** [comment]
It's already possible. You can run multiple Schedulers - each with LocalExecutor and it **should** work in principle.

**Reply by raphaelauv** [reply]
but as you explained there -> https://github.com/apache/airflow/discussions/26564#discussioncomment-3702194

it would no more be possible to scale the workers and the schedulers independently .

so my question is more : would it be a good idea to get ride of celery for **airflow workers** so they "act" the same way than **airflow triggerers** 

thanks

**Reply by potiuk** [reply]
> so my question is more : would it be a good idea to get ride of celery for airflow workers so they "act" the same way than airflow triggerers

I think we do not want to do it. IMHO Celery is precisely the solution we would have to implement ourselves if we want to implement the feature of "having one or multiple places where we decide which task to schedule" handled by "any number of workers that can handle it". 

Celery handles surprisingly a lot of complexity for us - being able ot submit a task to the queue to execute and basically stop worrying about it is precisely what Celery is all about. 

It handles:

* picking tasks from a queue
*  acknowledgment of the tasks when picked
*  handling forking/pre-forking of multiple processes that are picking the tasks
*  automatically restarting those processes when they fail
*  signal propagation - putting the processes in "offline" mode to let them finish the job
*  have flower UI to monitor those tasks run

If we want to build a distributed task execution mechanism, where we would like to run any number of workers to execute tasks then, welll, we would have to reimplement Celery essentially.
 
I doubt we can do better than them :).
